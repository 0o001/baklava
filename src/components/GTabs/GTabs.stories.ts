// @ts-nocheck
import { storiesOf } from '@storybook/vue';
import { withKnobs, object, text, boolean } from '@storybook/addon-knobs';
import GTabs from './GTabs.vue';
import GIcon from '../GIcon/GIcon.vue';
import GBox from '../GBox/GBox.vue';

const stories = storiesOf('GTabs', module);
stories.addDecorator(withKnobs);

stories
  .addParameters({ component: GTabs })
  .add('base', () => ({
    components: { GTabs, GBox },
    props: {
      tabs: {
        default: object('Tabs', [{
          type: 'all',
          label: 'tümü',
          totalElements: 12,
        }, {
          type: 'academy',
          label: 'akademi',
          totalElements: 486,
        }]),
      },
      type: {
        default: text('Type', 'all'),
      },
      showCount: {
        default: boolean('Show Count', false),
      },
    },
    methods: {
      handleTabClick (type) {
        this.type = type;
      },
    },
    template: `
      <GBox class="g-p-10 g-bg-bg-grey-100">
        <GTabs
          :tabs="tabs"
          :type="type"
          :showCount="showCount"
          @handleTabClick="handleTabClick"
        />
      </GBox>`,
  }))
  .add('new tab', () => ({
    components: { GTabs, GBox },
    props: {
      tabs: {
        default: object('Tabs', [{
          type: 'all',
          label: 'tümü',
        }, {
          type: 'academy',
          label: 'akademi',
          isNew: {
            color: 'white',
            text: 'YENİ',
            bgColor: 'red-500',
          },
        }]),
      },
      type: {
        default: text('Type', 'all'),
      },
      showCount: {
        default: boolean('Show Count', false),
      },
    },
    methods: {
      handleTabClick (type) {
        this.type = type;
      },
    },
    template: `
      <GBox class="g-p-10 g-bg-bg-grey-100">
      <GTabs
        :tabs="tabs"
        :type="type"
        :showCount="showCount"
        @handleTabClick="handleTabClick"
      />
      </GBox>`,
  }))
  .add('fluid', () => ({
    components: { GTabs },
    props: {
      tabs: {
        default: object('Tabs', [{
          type: 'all',
          label: 'tümü',
          totalElements: 11231232,
        }, {
          type: 'academy',
          label: 'akademi',
          totalElements: 0,
          tooltipText: 'selam',
        }]),
      },
      type: {
        default: text('Type', 'all'),
      },
      showCount: {
        default: boolean('Show Count', true),
      },
      fluid: {
        default: boolean('Fluid', true),
      },
    },
    methods: {
      handleTabClick (type) {
        this.type = type;
      },
    },
    template: `
      <GTabs
        :fluid="fluid"
        :tabs="tabs"
        :type="type"
        :showCount="showCount"
        @handleTabClick="handleTabClick"
      >
      <template slot="totalElementText">SORU</template>
</GTabs>`,
  })).add('slots', () => ({
    components: { GTabs, GBox, GIcon },
    props: {
      tabs: {
        default: object('Tabs', [{
          type: 'all',
          label: 'tümü',
          totalElements: 12,
        }, {
          type: 'academy',
          label: 'akademi',
          totalElements: 486,
        }]),
      },
      type: {
        default: text('Type', 'all'),
      },
      showCount: {
        default: boolean('Show Count', false),
      },
    },
    computed: {
      slotName () {
        return 'left|academy';
      },
    },
    methods: {
      handleTabClick (type) {
        this.type = type;
      },
    },
    template: `
      <GBox class="g-p-10 g-bg-bg-grey-100">
        <GTabs
          :tabs="tabs"
          :type="type"
          :showCount="showCount"
          @handleTabClick="handleTabClick"
        >
          <template v-slot:all="{ tab }">
            {{ tab.label }} {{ tab.isHovered ? 'Hovered' : '' }}
          </template>
          <template v-slot:[slotName]>
            <GIcon name="database" width="14"></GIcon>
          </template>
        </GTabs>
      </GBox>`,
  }));
