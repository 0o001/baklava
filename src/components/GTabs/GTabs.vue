<template>
  <div :class="classNames">
    <div
      class="tab"
      v-for="tab in tabs"
      :key="tab.type"
      :class="[{ 'active': tab.type === type }, tab.type]"
      @click="tab.type !== type && handleTabClick(tab.type)"
      @mouseenter="onMouseEnter(tab)"
      @mouseleave="onMouseLeave(tab)"
    >
      <slot
        :name="tab.type"
        :tab="{
          ...tab,
          isActive: tab.type === type,
          isHovered: tab.type === hoveredTabKey
        }"
      >
        <div class="tab-title">
          <GText
            class="tab-name g-d-flex g-d-align-center g-d-justify-center"
            variant="subtitle-03"
            :color="titleColor"
          >
            <slot
              :name="`left|${tab.type}`"
              :tab="{
                ...tab,
                isActive: tab.type === type,
                isHovered: tab.type === hoveredTabKey
              }"
            />
            <span :class="{ 'g-ml-4': !!$slots[`left|${tab.type}`] }">
              {{ label(tab.label) }}
            </span>
            <GTooltip
              v-if="tab.tooltipText"
              placement="bottom"
            >
              <template #tooltip-trigger>
                <GIcon
                  class="g-ml-4"
                  name="alert-circle"
                  size="14px"
                />
              </template>
              <template #tooltip-text>
                {{ label(tab.tooltipText) }}
              </template>
            </GTooltip>
            <GText
              v-if="tab.isNew"
              variant="small"
              :color="tab.isNew.color"
              :class="&quot;g-bg-&quot; + tab.isNew.bgColor"
              class="g-rad-sm g-py-4 g-px-10 g-ml-10 is-new"
            >
              {{ tab.isNew.text }}
            </GText>
          </GText>
        </div>
        <slot
          name="sub-title"
          :tab="{
            ...tab,
            isActive: tab.type === type,
            isHovered: tab.type === hoveredTabKey
          }"
        >
          <div
            v-if="showCount && tab.totalElements > -1"
            class="tab-count"
          >
            <GText
              variant="small"
              :color="subTitleColor"
            >
              {{ tab.totalElements | formatted }}
              <slot name="totalElementText">
                BİLDİRİM
              </slot>
            </GText>
          </div>
        </slot>
        <div class="seperator" />
      </slot>
    </div>
  </div>
</template>

<script lang="ts">

import GText from '../GText/GText.vue';
import GTooltip from '../GTooltip/GTooltip.vue';
import GIcon from '../GIcon/GIcon.vue';

export default {
  name: 'Tabs',
  components: {
    GTooltip,
    GText,
    GIcon,
  },
  props: {
    tabs: {
      type: Array,
      default: () => [],
    },
    type: {
      type: String,
      default: '',
    },
    showCount: {
      type: Boolean,
      default: false,
    },
    fluid: {
      type: Boolean,
      default: false,
    },
    titleColor: {
      type: String,
      default: () => 'main-grey-500',
    },
    subTitleColor: {
      type: String,
      default: () => 'main-grey-500',
    },
  },
  data () {
    return {
      hoveredTabKey: '',
    };
  },

  filters: {
    formatted (val:any) {
      return val.toString().replace(/(\d)(?=(\d{3})+(?!\d))/g, '$1.');
    },
  },
  computed: {
    classNames () {
      return {
        'g-tabs': true,
        '-fluid': this.fluid,
      };
    },
  },
  methods: {
    onMouseEnter (tab: any): void {
      this.hoveredTabKey = tab.type;
    },
    onMouseLeave (tab: any): void {
      this.hoveredTabKey = '';
    },
    handleTabClick (type: string): void {
      this.$emit('handleTabClick', type);
    },
    label (label) {
      return label
        ? label
        : '';
    },
  },
};
</script>

<style lang="scss" scoped>
.g-tabs {
  display: flex;
  flex-direction: row;
  border-bottom: 2px solid var(--bg-grey-500);
  background-color: var(--white);
  border-top-left-radius: var(--radius-md);
  border-top-right-radius: var(--radius-md);

  .tab {
    position: relative;
    text-align: center;
    padding: 16px;
    margin-bottom: -2px;
    cursor: pointer;
    border-bottom: 2px solid var(--bg-grey-500);

    &:last-child {
      .seperator {
        display: none;
      }
    }

    .tab-name{
      height: 20px;
    }

    .seperator {
      height: 16px;
      border-right: 1px solid var(--light-grey-800);;
      position: absolute;
      top: 18px;
      right: 0;
    }

    .tab-count {
      margin-top: 3px;
    }

    &.active {
      border-bottom: 2px solid var(--orange-500);

      .tab-title {
        .tab-name{
          color: var(--orange-500);
        }
      }

      .tab-count {
        .g-text {
          color: var(--main-grey-500);
        }
      }
    }

    &:hover {
      .tab-title {
        .tab-name{
          color: var(--orange-500);
        }
      }

      .tab-count {
        .g-text {
          color: var(--main-grey-500);
        }
      }
    }
    &:first-child{
      padding-left: 24px;
    }
  }

  &.-fluid {
    width: 100%;

    .tab {
      flex: 1;
    }
  }
}
</style>
