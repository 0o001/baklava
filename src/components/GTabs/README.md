# Tabs

## Slots

| Name               | Description |
| ------------------ | ----------- |
| `sub-title`        |             |
| `totalElementText` | &nbsp;      |

## Props

| Name         | Type      | Description | Default    |
| ------------ | --------- | ----------- | ---------- |
| `tabs`       | `Array`   |             | `() => []` |
| `type`       | `String`  |             | `""`       |
| `show-count` | `Boolean` |             | `false`    |
| `fluid`      | `Boolean` |             | `false`    |

## Computed Properties

| Name         | Type     | Description               |
| ------------ | -------- | ------------------------- |
| `classNames` | `object` | **Dependencies:** `fluid` |

## Events

| Name             | Description                                              |
| ---------------- | -------------------------------------------------------- |
| `handleTabClick` | <br/>**Arguments**<br/><ul><li>**`type: any`**</li></ul> |

## Methods

### handleTabClick()

**Syntax**

```typescript
handleTabClick(type: unknown): void
```

### label()

**Syntax**

```typescript
label(label: unknown): unknown
```

