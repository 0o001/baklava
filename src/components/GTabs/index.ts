import GTabs from './GTabs.vue';

export default GTabs;
