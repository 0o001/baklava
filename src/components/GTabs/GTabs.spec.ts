import { mount } from '@vue/test-utils';
import GTabs from './GTabs.vue';

describe('GTabs', () => {
  let wrapper: any;

  beforeEach(() => {
    wrapper = mount(GTabs);
  });

  afterEach(() => {
    jest.resetAllMocks();
    jest.restoreAllMocks();
  });

  it('should match snapshot given props ', async () => {
    wrapper.setProps({
      tabs: [{
        type: 'all',
        label: 'TÜMÜ',
        totalElements: 99,
      }, {
        type: 'announcement',
        label: 'DUYURU',
        totalElements: 12,
      }],
      type: 'all',
    });

    await wrapper.vm.$nextTick();
    expect(wrapper.element).toMatchSnapshot();
  });

  it('should tabs return [] with default props', () => {
    expect(wrapper.vm.tabs).toEqual([]);
  });

  it('should render GTabs Component', async () => {
    wrapper.setProps({
      tabs: [{
        type: 'all',
        label: 'TÜMÜ',
        totalElements: 99,
      }, {
        type: 'announcement',
        label: 'DUYURU',
        totalElements: 12,
      }],
      type: 'announcement',
    });

    await wrapper.vm.$nextTick();

    expect(wrapper.find('.tab').exists()).toBe(true);
    expect(wrapper.find('.active').exists()).toBe(true);
    expect(wrapper.find('.active .tab-title').text()).toBe('DUYURU');
  });

  it('should render GTabs Component when type not given', async () => {
    wrapper.setProps({
      tabs: [{
        type: 'all',
        label: 'TÜMÜ',
        totalElements: 99,
      }, {
        type: 'announcement',
        label: 'DUYURU',
        totalElements: 12,
      }],
    });

    await wrapper.vm.$nextTick();

    expect(wrapper.find('.tab').exists()).toBe(true);
    expect(wrapper.find('.active').exists()).toBe(false);
  });

  it('should render GTabs Component when showCount true', async () => {
    wrapper.setProps({
      tabs: [{
        type: 'all',
        label: 'TÜMÜ',
        totalElements: 99,
      }, {
        type: 'announcement',
        label: 'DUYURU',
        totalElements: 12,
      }],
      type: 'announcement',
      showCount: true,
    });

    await wrapper.vm.$nextTick();

    expect(wrapper.find('.tab').exists()).toBe(true);
    expect(wrapper.find('.active').exists()).toBe(true);
    expect(wrapper.find('.tab-count').exists()).toBe(true);
  });

  it('should render GTabs Component when fluid true', async () => {
    wrapper.setProps({
      tabs: [{
        type: 'all',
        label: 'TÜMÜ',
        totalElements: 99,
      }, {
        type: 'announcement',
        label: 'DUYURU',
        totalElements: 12,
      }],
      type: 'announcement',
      showCount: true,
      fluid: true,
    });

    await wrapper.vm.$nextTick();

    expect(wrapper.find('.-fluid').exists()).toBe(true);
    expect(wrapper.find('.tab').exists()).toBe(true);
    expect(wrapper.find('.active').exists()).toBe(true);
    expect(wrapper.find('.tab-count').exists()).toBe(true);
  });

  it('should change active tab', async () => {
    wrapper.setProps({
      tabs: [{
        type: 'all',
        label: 'TÜMÜ',
        totalElements: 99,
      }, {
        type: 'announcement',
        label: 'DUYURU',
        totalElements: 12,
      }],
      type: 'announcement',
    });

    await wrapper.vm.$nextTick();

    wrapper.findAll('.tab').at(0).trigger('click');
    wrapper.setProps({ type: 'all' });
    await wrapper.vm.$nextTick();

    expect(wrapper.find('.active .tab-title').text()).toBe('TÜMÜ');
  });

  it('should show new text if tab has a isNew property', async () => {
    wrapper.setProps({
      tabs: [{
        type: 'all',
        label: 'TÜMÜ',
        totalElements: 99,
      }, {
        type: 'announcement',
        label: 'DUYURU',
        totalElements: 12,
        isNew:{
          color: 'white',
          text: 'YENİ',
          bgColor: 'red-500',
        }
      }],
      type: 'announcement',
    });

    await wrapper.vm.$nextTick();

    expect(wrapper.find('.tab-name .is-new').text()).toBe('YENİ');
  });
});
