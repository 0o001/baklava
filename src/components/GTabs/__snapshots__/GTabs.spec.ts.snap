// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GTabs should match snapshot given props  1`] = `
<div
  class="g-tabs"
>
  <div
    class="tab active all"
  >
    <div
      class="tab-title"
    >
      <div
        class="tab-name g-d-flex g-d-align-center g-d-justify-center g-text-main-grey-500 -subtitle-03 g-text"
      >
         
        <span
          class=""
        >
          
            TÜMÜ
          
        </span>
         
        <!---->
         
        <!---->
      </div>
    </div>
     
    <!---->
     
    <div
      class="seperator"
    />
  </div>
  <div
    class="tab announcement"
  >
    <div
      class="tab-title"
    >
      <div
        class="tab-name g-d-flex g-d-align-center g-d-justify-center g-text-main-grey-500 -subtitle-03 g-text"
      >
         
        <span
          class=""
        >
          
            DUYURU
          
        </span>
         
        <!---->
         
        <!---->
      </div>
    </div>
     
    <!---->
     
    <div
      class="seperator"
    />
  </div>
</div>
`;
